src/mysqlsh_helpers.py

```diff
@@ -311,6 +369,197 @@ def add_instance_to_cluster(self, instance_address, instance_unit_label) -> None
+                f"shell.connect('{self.cluster_admin_user}:{self.cluster_admin_password}@{primary_address}')",
+                f"session.run_sql(\"UPDATE mysql.juju_units_operations SET executor='{unit_label}', status='in-progress' WHERE task='unit-teardown' AND executor='';\")",
+                f"acquired_lock = session.run_sql(\"SELECT count(*) FROM mysql.juju_units_operations WHERE task='unit-teardown' AND executor='{unit_label}';\").fetch_one()[0]",
+                "print(f'<ACQUIRED_LOCK>{acquired_lock}</ACQUIRED_LOCK>')",
+            )
+
+            output = self._run_mysqlsh_script("\n".join(acquire_lock_commands))
+            matches = re.search(r"<ACQUIRED_LOCK>(\d)</ACQUIRED_LOCK>", output)
+            if not matches:
+                raise MySQLRemoveInstanceError("Unable to acquire lock to remove unit")
+
+            # If unable to acquire the lock, raise an error to retry
+            acquired_lock = int(matches.group(1))
+            if not acquired_lock:
+                raise MySQLRemoveInstanceError("Did not acquire lock to remove unit")
+
+            # Get remaining cluster member addresses before calling mysqlsh.remove_instance()
+            remaining_cluster_member_addresses, valid = self._get_cluster_member_addresses(
+                exclude_unit_labels=[unit_label]
+            )
+            if not valid:
+                raise MySQLRemoveInstanceError("Unable to retrieve cluster member addresses")
+
+            # Remove instance from cluster, or dissolve cluster if no other members remain
+            logger.debug(
+                f"Removing instance {self.instance_address} from cluster {self.cluster_name}"
+            )
+            remove_instance_options = {
+                "password": self.cluster_admin_password,
+                "force": "true",
+            }
+            dissolve_cluster_options = {
+                "force": "true",
+            }
+            remove_instance_commands = (
+                f"shell.connect('{self.cluster_admin_user}:{self.cluster_admin_password}@{self.instance_address}')",
+                f"cluster = dba.get_cluster('{self.cluster_name}')",
+                "number_cluster_members = len(cluster.status()['defaultReplicaSet']['topology'])",
+                f"cluster.remove_instance('{self.cluster_admin_user}@{self.instance_address}', {json.dumps(remove_instance_options)}) if number_cluster_members > 1 else cluster.dissolve({json.dumps(dissolve_cluster_options)})",
+            )
+            self._run_mysqlsh_script("\n".join(remove_instance_commands))
+
+            # There is no need to release the lock if cluster was dissolved
+            if not remaining_cluster_member_addresses:
+                return
+
+            # Retrieve the cluster primary's address again (in case the old primary is scaled down)
+            # Release the lock by making a request to this primary member's address
+            primary_address = self._get_cluster_primary_address(
```

Comment:
should it have some retry logic in case this calls returns None?

Reply:
sync'ed offline - we should not be retrying any operations after the successful removal of the instance from the cluster. moved the block following the instance removal out of the `try/except`, and added a condition to raise an exception if the retrieval of the primary address failed in 0a55b4d219288fca33ec9f0673995488e198f445